File: templates/Datastructure/Fenwick2D.cpp
template<typename T>
  class Fenwick2D { // 1-indexed
    T n, m;
    vector<vector<T>> tree;
  public:
    Fenwick2D(T n, T m) : n(n), m(m) {
      tree.resize(n + 1, vector<T>(m + 1, 0));
    }
    void update(T x, T y, T delta) {
      while (x <= n) {
        T y1 = y;
        while (y1 <= m) {
          tree[x][y1] += delta;
          y1 += y1 & -y1;
        }
        x += x & -x;
      }
    }
  
    T query(T x, T y) {
      T result = 0;
      while (x > 0) {
        T y1 = y;
        while (y1 > 0) {
          result += tree[x][y1];
          y1 -= y1 & -y1;
        }
        x -= x & -x;
      }
      return result;
    }
  
    T rectangle_sum(T x1, T y1, T x2, T y2) {
      return query(x2, y2) - query(x1 - 1, y2) - query(x2, y1 - 1) + query(x1 - 1, y1 - 1);
    }
  };

File: python/DataStructure/Fenwick2D.py
class Fenwick2D:
  def __init__(self, n, m):
    self.n, self.m = n, m
    self.tree = [[0] * (m + 1) for _ in range(n + 1)]

  def update(self, x, y, delta):
    xi = x
    while xi <= self.n:
      yi = y
      while yi <= self.m:
        self.tree[xi][yi] += delta
        yi += yi & -yi
      xi += xi & -xi

  def query(self, x, y):
    result, xi = 0, x
    while xi > 0:
      yi = y
      while yi > 0:
        result += self.tree[xi][yi]
        yi -= yi & -yi
      xi -= xi & -xi
    return result

  def rectangle_sum(self, x1, y1, x2, y2):
    return (self.query(x2, y2) - self.query(x1 - 1, y2) - self.query(x2, y1 - 1) + self.query(x1 - 1, y1 - 1))